### PR TITLE
start testing the controller - found a bug :) where i was keeping the…

### PR DESCRIPTION
**Pull request by mick-warehime**
… keys by reference, probably not an issue with the actual pygame module, but did make it hard to debug failing tests. just copy the list each time

100% coverage

### CODE REVIEW

File: src/test/test_controller.py [hunk 1]
```diff
@@ -1,11 +1,138 @@
...
+        self.e = e
+
+    def test_bind(self) -> None:
+        ctrl = controller.Controller()
+
+        test_string = 'set a'
+        ctrl.bind(self.a_key, lambda: self.set_a(test_string))
+
+        pg.key.pressed[self.a_key] = 1
+
```
Comment (dvirk) [inline src/test/test_controller.py]: I would add self.assertEqual(self.a, '') before the call to update as well. And check that ctrl.bindings is a dictionary with one element.

File: src/test/test_controller.py [hunk 2]
```diff
@@ -1,11 +1,138 @@
...
+    def set_e(self, e: str) -> None:
+        self.e = e
+
+    def test_bind(self) -> None:
+        ctrl = controller.Controller()
+
+        test_string = 'set a'
+        ctrl.bind(self.a_key, lambda: self.set_a(test_string))
+
+        pg.key.pressed[self.a_key] = 1
```
Comment (dvirk) [inline src/test/test_controller.py]: why is the a one and note a bool? Just to match pygame?

Comment (mick-warehime) [reply src/test/test_controller.py]: pygame convention

File: src/test/test_controller.py [hunk 3]
```diff
@@ -1,11 +1,147 @@
...
+        ctrl.bind(self.c_key, lambda: self.set_c(test_string_c))
+        ctrl.bind_mouse(self.a_key, lambda: self.set_d(test_mouse_1))
+        ctrl.bind_mouse(self.b_key, lambda: self.set_e(test_mouse_2))
+
+        self.assertEqual(len(ctrl.bindings), 3)
+        self.assertEqual(len(ctrl.mouse_bindings), 2)
+
+        # every combination of 3 keys and two mouse presses
+        off_on = [0,1]
+        for key_1 in off_on:
```
Comment (dvirk) [inline src/test/test_controller.py]: how about the following to replace the nested for loops:
for key_1, key_2, key_3, mouse_1, mouse_2 in itertools.product(*(off_on * 5)):

File: src/test/test_controller.py [hunk 4]
```diff
@@ -1,11 +1,147 @@
...
+
+pg = pygame_mock.Pygame()
+
+sys.modules['pygame'] = pg
 import controller
 
 
 class ControllerTest(unittest.TestCase):
-    def test_set_functions(self) -> None:
+    def setUp(self) -> None:
```
Comment (dvirk) [inline src/test/test_controller.py]: just to confirm: This is called before every test?

Comment (mick-warehime) [reply src/test/test_controller.py]: called once before each test
https://stackoverflow.com/questions/23667610/difference-between-setupclass-and-setup-in-python-unittest

**Pull request merged**